[flare] add a permissions.log file into the flare archive

remeh opened the pull request:
### What does this PR do?

This PR adds a `permissions.log` file to the flare archive. As the files shipped in the flare archive are contained in a zip archive, we lost their permissions info of the original system. This file provide them.

### Motivation

The `permissions.log` file was available in the agent v5 flare archive but not in the v6.
It could be used to diagnose permissions errors of the agent setup.

### Additional Notes

There is a slight difference in the file format, the mode info were displayed as an octal number (e.g. `0644`) while this version displays them like a `ls -l` output: `-rw-r--r--`
The complexity to display them as `0644` _with the file type_ was ridiculous compared to calling the method for the latter display. Or maybe I've missed an obvious solution (see https://golang.org/src/os/types.go?#L65 for reference).

Note that this PR doesn't cover the Windows build.

remeh committed:
[flare] implement methods to add permissions info of files added to the flare archive.

Following the format of the permissions.log generated in the Agent 5
except for the mode info which looks like an `ls -l` call:
`-rw-r--r--` instead of `0644` in the old format.

remeh committed:
[flare] creates the permissions.log file in flare while creating the archive.

Only for non-windows system, windows generated flares still won't have
any `permissions.log` file.

remeh committed:
[flare] unit tests for the permissions.log file writing.

remeh committed:
[api] expose the auth token filepath.

remeh committed:
[flare] add auth token permissions infos.

remeh committed:
Add release note entry.

remeh committed:
[flare] fix perms file method signature for windows build.

truthbk commented on pkg/flare/archive.go:
```diff
@@ -66,6 +68,11 @@ func createArchive(zipFilePath string, local bool, confSearchPaths SearchPaths,
 	}
 
+	// create the permissions.log file and write its headers
+	if err := initPermsInfo(tempDir, hostname, os.ModePerm); err != nil {
```
I like the interface, but I think I may prefer using a map to store the filename, perms - we could add those with `addPermsInfo` function - and then committing to the file in a single operation as opposed to open/write, etc for every file we add to `permissions.log` what do you think?

remeh replied:
You're right, I was not so happy with the open/write solution but I didn't want to move an opened `*File` in the `createArchive`, moreover, my first shot was with a dedicated data struct and I didn't choose it because it needs to change some existing func signatures, but probably worth it. I'll implement the changes.

truthbk commented on pkg/flare/archive_nix.go:
```diff
@@ -7,10 +7,83 @@
 }
+
+// initPermsInfo creates the permissions.log file and write the headers.
+func initPermsInfo(tempDir, hostname string, p os.FileMode) error {
```
As I mentioned above, I would probably use a struct, and use this to initialize a map for our file permissions, and destination file, then use the `addPermsInfo` to populate the map. And finally some `commitPermsInfo()` function or similar to write all that info to the file. What do you think?

remeh replied:
Yes the `commit` solution looks like the good one. Although, I'm wondering if we should invalidate the "permissions struct" if something wrong happened during the stat parts, or going with a best effort and write as much permissions info as we can (what we were doing in the current implementation).

truthbk commented on pkg/flare/archive_nix.go:
```diff
@@ -7,10 +7,83 @@
+
+	u, err := user.LookupId(strconv.Itoa(int(sys.Uid)))
+	if err != nil {
+		return fmt.Errorf("can't lookup for uid info: %v", err)
```
Should we return and not write in these scenarios, or still write (`"???"` as above), log something, and then return an error? These failures I hope will be rare in most conditions, wondering if a best-effort is preferable.... 🤔

truthbk commented on pkg/flare/archive_nix_test.go:
```diff
@@ -0,0 +1,81 @@
+)
+
+// Ensure the permissions.log is being created
+func TestPermsFile(t *testing.T) {
```
🍰

truthbk reviewed (commented):
Nice implementation, great stuff! I have a couple of questions, and just an idea about writing a single time to `permissions.log` as opposed to multiple times (one per permission logged). Let me know what you think!

remeh committed:
[flare] refactor using a state during the archive creation instead of writing each time to the file.

In order to not open the permissions.log file for every lines. I've used
a new struct to store filepaths that will be added to the
`permissions.log` file. Before finally creating the .zip archive, we create
the `permissions.log` file with all files information in one shot.

Slightly change the signature of two methods shared with the DCA:
  - `zipConfigFiles`
  - `zipLogFiles`

This commit doesn't add the `permissions.log` file for the DCA flare archive,
this is why the nil pointer is handled in those methods (to avoid the DCA to be
forced to provide a *PermsInfos instance).

remeh committed:
[flare] fix a bug about reading the copied file instead of the original one.

remeh committed:
[flare] add a missing err test while writing permissions.log header

remeh committed:
[flare] PermsInfos func should not be public.

We don't need them from outside of this pkg.

truthbk commented on pkg/flare/archive.go:
```diff
@@ -33,6 +34,36 @@ import (
+// The filemode field is the file mode which will be used to create
+// the temporary permissions.log file on filesystem before being written in the zip.
+type PermsInfos struct {
+	tempDir  string
```
I think I would remove `tempDir` from the structure and just pass the relevant info to the commit/write functions? Like, commit could take the path to the destination file or something along those lines.

remeh replied:
Ok! Not a problem and makes sense providing this info during the commit phase.

truthbk commented on pkg/flare/archive.go:
```diff
@@ -33,6 +34,36 @@ import (
+	infos map[string]permsInfo
+}
+
+type permsInfo struct {
```
nit: could we rename this into `filePermsInfo`?

remeh replied:
Definitely. Would you also prefer File*s*PermsInfos for the public struct containing all infos?

truthbk replied:
I don't think you necessarily need to change the name to `FilesPermsInfos()` (though you can if you want, your decision. I didn't mention changing it because the current name leaves room for _other_ permissions maybe unrelated to files - though nothing really comes to mind now of what that could be). Regarding making it private, I think that might make sense if we're not using it anywhere. We can always make it public in the future 👍

truthbk reviewed (commented):
Almost, there, left two annoying nitpicks you can probably fix in ~60 seconds 🙈

remeh committed:
[flare] fix permissions check for log files (was using the dst file).

remeh committed:
[flare] changed the PermsInfos composition and made it private.

Bonus, `permsInfo` becomes `filePermsInfo`.

truthbk reviewed (dismissed):
Looks good! Thank you @remeh 🙇

remeh committed:
[flare] archive dca flare generation also create a `permissions.log` file in the archive.

remeh committed:
[flare] wrong func assignement for windows stubbed funcs.

remeh committed:
[flare] while creating dca flare archive, do not fail on permissions.log creation error, just log the error.

truthbk reviewed (approved):
Let's merge this one now! 🚀
